Refactor ItineraryBody so it can be formatted like the original ItineraryBody in otp-react-redux

evansiroky opened the pull request:
This PR adds and refactors a bunch of stuff so that the ItineraryBody component can be styled and rendered as close as possible to the original ItineraryBody component in otp-react-redux. This also introduces some breaking changes to improve code maintenance and avoid loading unnecessary components.

- Creates stories for each of the mock itineraries with otp-rr styling
- Extracts some individual components into single files to better organize things
- Adds new props to the root component:
  - LineColumnContent
  - showLegIcon
  - showMapButtonColumn
  - TransitLegSubheader
- Places all default components for slots into the defaults folder
- BREAKING: does not automatically use the default components for each slot. It is now expected for the developer to include these on their own. Doing this allows the application to not have to load potentially unnecessary components and code.
- Adds meaningful arguments when calling the `frameLeg` function (fixes #94).

evansiroky commented:
@fpurcell, the null values occurring while clicking on the destination happen because there isn't another leg that continues after the destination. Maybe I'll include another parameter that indicates that the item clicked is the destination.

binh-dam-ibigroup commented:
Would it be possible to remove the custom font from the unstyled itinerary bodies?

binh-dam-ibigroup commented on packages/itinerary-body/src/AccessLegBody/access-leg-steps.js:
```diff
@@ -0,0 +1,39 @@
+export default function AccessLegSteps({ steps }) {
+  return (
+    <Styled.Steps>
+      {steps.map((step, k) => {
```
Can steps be null/undefined?

fpurcell commented:
> @fpurcell, the null values occurring while clicking on the destination happen because there isn't another leg that continues after the destination. Maybe I'll include another parameter that indicates that the item clicked is the destination.

Yeah something.  The clicking on that map icon in the itinerary is going to then be used to move the map to that location. So having a record with coords available in the frameLeg() callback would be necessary.

binh-dam-ibigroup commented on packages/itinerary-body/src/AccessLegBody/rented-vehicle-subheader.js:
```diff
@@ -0,0 +1,76 @@
+
+/**
+ * A component to display vehicle rental data. The word "Vehicle" has been used
+ * because a future refactor is intended to combine car rental, bike rental
```
"because we plan to..."

evansiroky replied:
Let's just wait until this refactor actually happens to change this comment.

binh-dam-ibigroup commented on packages/itinerary-body/src/AccessLegBody/rented-vehicle-subheader.js:
```diff
@@ -0,0 +1,76 @@
+/**
+ * A component to display vehicle rental data. The word "Vehicle" has been used
+ * because a future refactor is intended to combine car rental, bike rental
+ * and micromobility rental all within this component. The future refactor is
```
"We are assuming that..."
If it is a new feature, I would avoid using the word refactor.

binh-dam-ibigroup commented on packages/itinerary-body/src/AccessLegBody/rented-vehicle-subheader.js:
```diff
@@ -0,0 +1,76 @@
+  // e.g., Pick up REACHNOW rented car XYZNDB OR
+  //       Pick up SPIN E-scooter
+  //       Pick up shared bike
+  return <Styled.PlaceSubheader>{rentalDescription}</Styled.PlaceSubheader>;
```
Move this before the big if block.

evansiroky replied:
This rentalDescription variable is modified in the immediately preceding if block, so I don't understand what you are asking me to do.

binh-dam-ibigroup replied:
I was suggesting to move the comment, so it describes the big if block does.
Alternatively, you can state in the existing comment that the examples are produced by the if block above.

binh-dam-ibigroup replied:
Well, never mind, that is probably not necessary anymore as the block above has shrunk...

binh-dam-ibigroup commented on packages/itinerary-body/src/AccessLegBody/rented-vehicle-subheader.js:
```diff
@@ -0,0 +1,76 @@
+  let rentalDescription = "Pick up";
+  if (leg.rentedBike) {
+    // TODO: Special case for TriMet may need to be refactored.
+    rentalDescription += ` shared bike`;
```
I would rather not concatenate strings (for future translations, if it gets there).

binh-dam-ibigroup commented on packages/itinerary-body/src/AccessLegBody/rented-vehicle-subheader.js:
```diff
@@ -0,0 +1,76 @@
+        leg.from.networks,
+        configCompanies
+      );
+      rentalDescription += ` ${companiesLabel}`;
```
Same here

binh-dam-ibigroup commented on packages/itinerary-body/src/AccessLegBody/rented-vehicle-subheader.js:
```diff
@@ -0,0 +1,76 @@
+      rentalDescription = "Continue using rental";
+    }
+
+    rentalDescription += ` ${modeString} ${vehicleName}`;
```
Same here.

binh-dam-ibigroup commented on packages/itinerary-body/src/TransitLegBody/alerts-body.js:
```diff
@@ -0,0 +1,54 @@
+              sameDay: `${timeFormat}, [Today]`,
+              nextDay: `${timeFormat}, [Tomorrow]`,
+              lastDay: `${timeFormat}, [Yesterday]`,
+              lastWeek: `${longDateFormat}`,
```
Do we need brackets around Today, Tomorrow, Yesterday?

binh-dam-ibigroup commented on packages/itinerary-body/src/defaults/line-column-content.js:
```diff
@@ -0,0 +1,58 @@
+}
+
+LineColumnContent.propTypes = {
+  /** whether this leg is an interlined-transit leg */
```
Capital W

randolphjones commented:
@fpurcell Thanks for pointing that out. Yes, we are trying to frame the selected leg on the map.
Two ways we can implement this: We can get the entire leg object as an argument and frame the view based on that or we can get the index and use that to select the leg from the user's active itinerary. Either way, we use the existing bound utilities with a leg object and frame up the view that way. Whether we get the entire leg from the get-go or the index of the leg... we can be flexible on that.

Taking a glance at the code in Evan's earlier commits, looks like we get both? I think we can definitely work w that 👍

binh-dam-ibigroup commented on packages/itinerary-body/src/defaults/route-description.js:
```diff
@@ -0,0 +1,34 @@
+        {headsign && (
+          <span>
+            <Styled.LegDescriptionHeadsignPrefix>
+              {" to "}
```
I think you can remove `{"`

evansiroky replied:
The spaces are needed, otherwise, the words will get squished together.

binh-dam-ibigroup commented on packages/itinerary-body/src/index.js:
```diff
@@ -36,46 +40,24 @@ const ItineraryBody = ({
   const rows = [];
   let followsTransit = false;
+  let lastLeg;
   itinerary.legs.forEach((leg, i) => {
```
Can you use array.map to achieve this?

binh-dam-ibigroup commented:
What is the`otp-react-redux` subfolder for? Is it a mock folder?

evansiroky commented:
> Would it be possible to remove the custom font from the unstyled steps?
> 
> ![image](https://user-images.githubusercontent.com/56846598/77573426-0cb4f100-6ea7-11ea-9c8b-2e3fe3958777.png)

@binh-dam-ibigroup, there is no custom font present. The differences in font are due to default browser styling rules. It is generally expected that implementers will define fonts at a higher level than just this component.

evansiroky commented:
I believe I've addressed most of the comments here. @fpurcell, regarding your requests, here is what I have done:

> a) rewrite 'Ride 3 min / 2 stops ▼ | Trip Viewer' as '3-minute ride (2 stops) ▼'
(or 'Ride 3 min (2 stops) ▼' ... btw, bold used here to differentiate desired to be state)

See story titled `ItineraryBody with walk-transit-walk itinerary with custom view trip button activated and custom route abbreviation` and corresponding code.

> b) remove the MA from the bubble

See story titled `ItineraryBody with walk-transit-walk itinerary with custom view trip button activated and custom route abbreviation` and corresponding code.

> c) align the time near the leg circle

See story titled `Styled ItineraryBody with walk-transit-walk itinerary` and corresponding code.

> frameLeg() callback

The frameLeg callback now also sends the `isDestination` and `place` values in callbacks.

binh-dam-ibigroup commented:
This case producing a validation error. Is that intended?
![image](https://user-images.githubusercontent.com/56846598/77691458-53741b00-6f7b-11ea-9f9f-6c54e8e6e16c.png)

evansiroky commented:
@binh-dam-ibigroup, see latest commit.
